woocommerce/google-listings-and-ads
# API PULL - Prevent notification errors for deleted products

### Changes proposed in this Pull Request:

When we send notifications we take as granted that the product or coupon are available in DB when handling its metas. This is not always true as the Job and some intermediate functions can run after the Product or Coupon is permanently deleted. 

When this happens, the notifications meta handler or some other functions will throw some exceptions. 

This PR fixes that ensuring that the item is available. Introducing the next changes:

- Check the item is available on Update and Create Topics. Boil in case it isn't. 
- Run the delete Notification right away (without AS job) when the product doesn't exist.
- Consider NON-existent products as already Notified products. This could trigger (very rarely, I wasn't able to reproduce it yet) unnecessarily DELETE notifications when we permanently delete a product that was never notified. But I think even if this may happen it would be better than missing a needed notification. 


### Detailed test instructions:


1. Follow the steps in https://github.com/woocommerce/google-listings-and-ads/wiki/API-PULL-Feature-(Beta)#enabling-the-feature and https://github.com/woocommerce/google-listings-and-ads/wiki/API-PULL-Feature-(Beta)#enabling-the-feature for enabling and granting access to Notifications feature.
2. Create a Product - See the `gla/jobs/notifications/products/process_item` job is scheduled with CREATE topic. 
3. Run it. See the Notification happens in the Logs
4. Update the product - See the `gla/jobs/notifications/products/process_item` job is scheduled with UPDATE topic.
5. Run it. See the Notification happens in the Logs
6. Trash the product - See the `gla/jobs/notifications/products/process_item` job is scheduled with DELETE topic.
7. Run it. See the Notification happens in the Logs
8. Create a Product - See the `gla/jobs/notifications/products/process_item` job is scheduled with CREATE topic. 
9. Don't run it.
10. Delete the product via `wp_delete_post( {post_id}, true );`
11. Run the job from step 8. See the Notification DOESNT happen in the Logs. NO exception happens.
12. Repeat steps 2-4. Don't run the UPDATE topic Job
13. Delete the product via `wp_delete_post( {post_id}, true );`
14. Run the update job from step. See the Notification DOESNT happen in the Logs. NO exception happens.
15. Repeat steps 2-3 and 6. Don't run the DELETE topic Job
16. Delete the product via `wp_delete_post( {post_id}, true );`
17. Now run the job. See the Notification YES happens in the Logs. NO exception happens.
18. Repeat all the steps for Coupons. 

### Additional details

I considered as well having some kind of table/options to keep the Notification metas so we don't depend anymore on the item itself. But that would require a big re-structure of the Notifications so this may be discussed for phase 2.
